src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -218,6 +219,12 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     // Call UpdateCompletedStreams() prior to frame processing in order to remove any streams that have exceeded their drain timeouts.
                     UpdateCompletedStreams();
 
+                    if (result.IsCanceled)
+                    {
+                        // Heartbeat will cancel ReadAsync and trigger expiring unused streams from pool.
+                        StreamPool.RemoveExpired(SystemClock.UtcNowTicks);
```

Comment:
Nit: UpdateCompletedStreams() also calls SystemClock.UtcNowTicks, so it'd be a little more efficient to store the value and a variable and pass it to both UpdateCompletedStreams() and RemoveExpired().

Reply:
I didn't because this runs infrequently.

Reply:
To be fair, SystemClock.UtcNowTicks is already cached every second, so it would only save a virtual call. That said, we're already retrieving it after every read for UpdateCompletedStreams(), so I don't see why we wouldn't reuse it.

Reply:
If we wanted, we could update the `UpdateCompletedStreams()` to only check for the drain timeout when there's a canceled ReadResult. That'd be a similar optimization to how you're only calling StreamPool.RemoveExpired() for a canceled ReadResult. Seems easy enough even though it wouldn't be a huge gain since we have to iterate over all the _completedStreams either way. Just a though.

Reply:
Off topic: Have we ever considered only running UpdateCompletedStreams on heartbeat?

Reply:
Yes. The problem is UpdateCompletedStreams does more than remove _completedStreams after a timeout. It's removing streams from the non-concurrent _streams Dictionary that were just added to _completedStreams from random ThreadPool threads after middleware completed.

In practice, most of these streams already have the EndStreamReceived flag set making them immediately eligible for removal from the Dictionary, but we cannot do so from a random ThreadPool tread. A ConcurrentDictionary could fix that, but that would lead to contention when adding streams to or retrieving streams from the ConcurrentDictionary.

If we only ran UpdateCompletedStreams once a second, a very large number of _completedStreams could build up before any get returned to the pool in high RPS scenarios. In low RPS scenarios, _completedStreams should stay very small which should make UpdateCompletedStreams relatively inexpensive.